CORDA-3899 Refactor flow's transient fields

lankydan opened the pull request:
Refactor `FlowStateMachineImpl.transientValues` and
`FlowStateMachineImpl.transientState` to stop the fields from exposing
the fact that they are nullable.

This is done by having private backing fields `transientValuesReference`
and `transientStateReference` that can be null. The nullability is still
needed due to serialisation and deserialisation of flow fibers. The
fields are transient and therefore will be null when reloaded from the
database.

Getters and setters hide the private field, allowing a non null field to
returned.

There is no point other than in `FlowCreator` where the transient fields
can be null. Therefore the non null checks that are being made are
valid.

lankydan committed:
CORDA-3899 Refactor flow's transient fields

Refactor `FlowStateMachineImpl.transientValues` and
`FlowStateMachineImpl.transientState` to stop the fields from exposing
the fact that they are nullable.

This is done by having private backing fields `transientValuesReference`
and `transientStateReference` that can be null. The nullability is still
needed due to serialisation and deserialisation of flow fibers. The
fields are transient and therefore will be null when reloaded from the
database.

Getters and setters hide the private field, allowing a non null field to
returned.

There is no point other than in `FlowCreator` where the transient fields
can be null. Therefore the non null checks that are being made are
valid.

lankydan committed:
CORDA-3899 Changes after rebase

lankydan committed:
CORDA-3899 Remove unused import

lankydan committed:
CORDA-3841 Tidy up code

lankydan committed:
CORDA-3841 Tidy up code

lankydan committed:
CORDA-3841 Use reflection in test to bypass check

williamvigorr3 reviewed (dismissed):
This makes things a lot cleaner.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt:
```diff
@@ -86,52 +85,57 @@ class FlowStateMachineImpl<R>(override val id: StateMachineRunId,
-    internal var transientValues: TransientReference<TransientValues>? = null
-    internal var transientState: TransientReference<StateMachineState>? = null
+    private var transientValuesReference: TransientReference<TransientValues>? = null
+    internal var transientValues: TransientValues
```
Safe as no backing field when get and set both overridden and no reference to `field`

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt:
```diff
@@ -86,52 +85,57 @@ class FlowStateMachineImpl<R>(override val id: StateMachineRunId,
+        }
+
+    private var transientStateReference: TransientReference<StateMachineState>? = null
+    internal var transientState: StateMachineState
```
Safe as no backing field when get and set both overridden and no reference to `field`

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt:
```diff
@@ -86,52 +85,57 @@ class FlowStateMachineImpl<R>(override val id: StateMachineRunId,
-        get() = transientState?.value?.senderUUID
+    override val logger = log
 
-    private fun <A> getTransientField(field: KProperty1<TransientValues, A>): A {
```
Oh.... hmmm.... I think this might have existed because the style `transientValues.field` in suspending methods ended up leaving a copy of `transientValues` on the stack sometime, depending on what kotlin did exactly.  Vaguely remember @exFalso discussing it.  So it might be you need to retain and use this instead.

I think we could perhaps prove once and for all by modifying `TransientValues` to make it have custom serializer methods (see https://github.com/EsotericSoftware/kryo#kryoserializable) such that it throws exceptions if read or written, and see what happens.  This seems prudent anyway in case we ever accidentally do the wrong thing.

lankydan replied:
Ok, if only there was a comment or commit stating that this is for some magic reason 😄

lankydan replied:
I added the custom serialization and didn't see any issues (on the test that I ran... which did a retry from its checkpoint).  Just to be sure, I took out the `@Transient` annotation in `TransientReference` and saw it hit my serialization functions. Then to be double sure I took out my functions and saw that broke by default when retrying a flow (so no custom serializer and no `@Transient` annotation).

exFalso commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowStateMachineImpl.kt:
```diff
@@ -476,7 +480,7 @@ class FlowStateMachineImpl<R>(override val id: StateMachineRunId,
     @Suspendable
     override fun <R : Any> suspend(ioRequest: FlowIORequest<R>, maySkipCheckpoint: Boolean): R {
-        val serializationContext = TransientReference(getTransientField(TransientValues::checkpointSerializationContext))
+        val serializationContext = TransientReference(transientValues.checkpointSerializationContext)
```
I'd recommend looking at the bytecode changes here. If there is no stack slot allocated for `transientValues` then it's all good. I vaguely remember that the only way I could hide any intermediate object was to have them as return values from functions and re-wrap them in a transient reference, but this needs verification. @rick-r3's suggestion also works wrt adding a serializer for `TransientValues`.

Also, props for touching this code hahaha

lankydan committed:
CORDA-3841 Ensure transient values/state is not serialized

Add custom kryo serialisation and deserialisation to `TransientValues`
and `StateMachineState` to ensure that neither of the objects are ever
touched by kryo.

lankydan commented:
Run tests: slow

lankydan commented:
Run tests: slow

lankydan commented:
Run tests: slow

lankydan commented:
Run tests: slow

lankydan commented:
Run tests: slow

lankydan commented:
Run tests: slow
